### PR TITLE
Keep the context that we are inside macro in nested macro

### LINKED ISSUE #2830: rustfmt  0.6.1-stable (49279d7 2018-05-08) produces not compilable for macros
I try run `cargo fmt` against Dushistov/rust-nmea@1de1c13569c557233ee7148a39b861860b61407e and as result `cargo build` failed after changes made by `cargo fmt`.

The problem in commas:

```Rust
named!(
    do_parse_gsv<GsvData>,
    map_res!(
        do_parse!(
            number_of_sentences: map_res!(digit, parse_num::<u16>)
                >> char!(',')
                >> sentence_index: map_res!(digit, parse_num::<u16>)
                >> char!(',')
                >> total_number_of_sats: map_res!(digit, parse_num::<u16>)
                >> char!(',')
                >> sat0: opt!(complete!(parse_gsv_sat_info))
                >> sat1: opt!(complete!(parse_gsv_sat_info))
                >> sat2: opt!(complete!(parse_gsv_sat_info))
                >> sat3: opt!(complete!(parse_gsv_sat_info))
                >> (
                    number_of_sentences,
                    sentence_index,
                    total_number_of_sats,
                    sat0,
                    sat1,
                    sat2,
                    sat3,
                )
        ),
        construct_gsv_data
    )
);
```

If I remove **,** after **sat3** everything will start working again.

I use rustfmt that comes with current stable rustc via rustup.

### LINKED ISSUE #2857: Changing vec!(...) to vec![...] breaks convert_args! macro
I originally filled a bug with the developer of the maplit crate: bluss/maplit#21 .

My issue is that rustfmt seems to change `convert_args!(vec!(1, 2, 3))` to `convert_args!(vec![1, 2, 3])` which doesn't compile.

The maplit developer replied with:

> Combining macros like this was deemed idiomatic (see nom, it prefers ()'s), so it seems like an issue with rustfmt.

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #2830.
Closes #2857.

### CODE REVIEW

File: src/macros.rs [hunk 1]
```diff
@@ -149,12 +149,12 @@ pub fn rewrite_macro(
     shape: Shape,
     position: MacroPosition,
 ) -> Option<String> {
-    context.inside_macro.replace(true);
+    let old_value = context.inside_macro.replace(true);
```
Comment (nrc) [inline src/macros.rs]: Could you use an RAII pattern here so there is no change of missing the `replace` at the end of the function?

Comment (topecongiro) [reply src/macros.rs]: Hmm, I am sorry, but I am not sure how to apply an RAII pattern here. Does `RefCell` provides such thing? Either way, I have tried to mimic the pattern here. Please let me know if this is not what you expected, or there is a better way to do this!

Comment (nrc) [reply src/macros.rs]: RAII in Rust would be something like:

```
struct InsideMacroRestore<'a> {
    context: &RewriteContext,
    old_value: bool,
}

impl<'a> InsideMacroRestore<'a> {
    // Call this as `let _imr = InsideMacroRestore(context, true);` which sets
    // the value and stores the old one.
    fn replace(context: &'a RewriteContext, new_value: bool) -> InsideMacroRestore<'a> {
        let old_value = context.inside_macro.replace(new_value);
        InsideMacroRestore { context, old_value }
    }
}

impl<'a> Drop for InsideMacroRestore<'a> {
    // Called automatically when `_imr` goes out of scope, which restores the old
    // value of inside_macro
    fn drop(&mut self) {
        self.context.inside_macro.replace(self.old_value)
    }
}
```

`MutexLock` is an example of such thing, although there the object is actually useful, here we just want the `Drop` impl

**Comment by topecongiro** [comment]
Added commits to use RAII pattern and fix #2857.

**Pull request merged**